pkg/ring/lifecycler.go

```diff
@@ -634,6 +864,19 @@ func (i *Lifecycler) changeState(ctx context.Context, state IngesterState) error
 	return i.updateConsul(ctx)
 }
 
+// getLastRing returns a copy of the last ring saved.
+func (i *Lifecycler) getLastRing() *Desc {
+	i.countersLock.Lock()
+	defer i.countersLock.Unlock()
+
+	if i.lastRing == nil {
+		return nil
+	}
+
+	r := *i.lastRing
```

Comment:
The result of this function gets modified by `joinIncrementalTransfer` and `leaveIncrementalTransfer` and read simultaneously in another goroutine (token checker). It'd be a race condition to not return a copy - I'll explicitly copy the whole structure.

Reply:
Then it would make more sense to me to create a copy where it's needed, and not here.

Reply:
It's safer to return a copy here and not at the call site. We don't want to allow callers to modify our copy of the last ring. I would say it's common practice to return copies in get methods rather than trusting callers to do the right thing.

Reply:
That is a safe thing to do when writing APIs for public consumption, but not always needed for internal usage, when you control all callers. For example token_checker.go doesn't need a copy. But it's ok.